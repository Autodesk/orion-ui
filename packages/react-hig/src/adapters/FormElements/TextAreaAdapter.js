/**
 Copyright 2016 Autodesk,Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */
import * as HIG from 'hig.web';

import HIGElement from '../../elements/HIGElement';
import * as PropTypes from 'prop-types';
import createComponent from '../createComponent';

class TextAreaAdapter extends HIGElement {
  constructor(initialProps) {
    super(HIG.TextArea, initialProps);
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    for (let i = 0; i < updatePayload.length; i += 2) {
      const propKey = updatePayload[i];
      const propValue = updatePayload[i + 1];

      switch (propKey) {
        case 'disabled': {
          propValue ? this.hig.disable() : this.hig.enable();
          break;
        }
        case 'instructions': {
          this.hig.setInstructions(propValue);
          break;
        }
        case 'label': {
          this.hig.setLabel(propValue);
          break;
        }
        case 'name': {
          this.hig.setName(propValue);
          break;
        }
        case 'placeholder': {
          this.hig.setPlaceholder(propValue);
          break;
        }
        case 'required': {
          propValue
            ? this.hig.required(propValue)
            : this.hig.noLongerRequired();
          break;
        }
        case 'value': {
          this.hig.setValue(propValue);
          break;
        }
        case 'onBlur': {
          const dispose = this._disposeFunctions.get('onBlurDispose');

          if (dispose) {
            dispose();
          }

          this._disposeFunctions.set('onBlur', this.hig.onBlur(propValue));
          break;
        }
        case 'onChange': {
          const dispose = this._disposeFunctions.get('onChangeDispose');

          if (dispose) {
            dispose();
          }

          this._disposeFunctions.set('onChange', this.hig.onChange(propValue));
          break;
        }
        case 'onFocus': {
          const dispose = this._disposeFunctions.get('onFocusDispose');

          if (dispose) {
            dispose();
          }

          this._disposeFunctions.set('onFocus', this.hig.onFocus(propValue));
          break;
        }
        case 'onInput': {
          const dispose = this._disposeFunctions.get('onInputDispose');

          if (dispose) {
            dispose();
          }

          this._disposeFunctions.set('onInput', this.hig.onInput(propValue));
          break;
        }
        default: {
          return;
        }
      }
    }
  }
}

const TextAreaComponent = createComponent(TextAreaAdapter);

TextAreaComponent.propTypes = {
  disabled: PropTypes.bool,
  instructions: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string,
  placeholder: PropTypes.string,
  required: PropTypes.string,
  value: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onInput: PropTypes.func
};

TextAreaComponent.__docgenInfo = {
  props: {
    disabled: {
      description: 'prevents interaction with the text field'
    },
    icon: {
      description: 'icon for the text field'
    },
    instructions: {
      description: 'instructional text for the text field'
    },
    label: {
      description: 'label for the text field'
    },
    name: {
      description: 'name of the field as submitted with a form'
    },
    onBlur: {
      description: 'callback called when user moves focus away from the text field'
    },
    onChange: {
      description: 'callback called when user changes the value of the text field and moves focus away'
    },
    onFocus: {
      description: 'callback called when user moves focus onto the text field'
    },
    onInput: {
      description: 'callback called when user changes the value of the text field'
    },
    placeholder: {
      description: 'text prompting the user to enter text'
    },
    required: {
      description: 'text indicating that the user must enter a value for this text field'
    }
  }
};

export default TextAreaComponent;
