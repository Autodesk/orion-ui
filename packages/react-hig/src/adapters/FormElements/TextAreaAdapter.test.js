/**
 Copyright 2016 Autodesk,Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */
import { mount } from 'enzyme';
import * as HIG from 'hig.web';
import React from 'react';

import TextAreaAdapter from './TextAreaAdapter';


const inputId = '1234';

describe('<TextAreaAdapter>', () => {

  function createHigTextArea(defaults = {}) {
    const higContainer = document.createElement('div');

    const higTextArea = new HIG.TextArea({ ...defaults });

    higTextArea.mount(higContainer);
    setLabelForInputId(higContainer);

    return { higTextArea, higContainer };
  }

  function createOrionTextArea(props) {
    return <TextAreaAdapter {...props} />;
  }

  function setLabelForInputId(higContainer) {
    // to adjust for the randomly generated id
    const label = higContainer.querySelector('label');
    const input = higContainer.querySelector('textarea');
    label.setAttribute('for', inputId);
    input.setAttribute('id', inputId);
  }


  it('renders a text field', () => {
    const defaults = { name: 'mySpecialField' };

    const { higTextArea, higContainer } = createHigTextArea(defaults);
    const container = document.createElement('div');
    const wrapper = mount(createOrionTextArea(defaults), {
      attachTo: container
    });
    setLabelForInputId(container);

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders a text field with initial props', () => {
    const defaults = {
      icon: 'assets',
      instructions: "Don't just do something, sit there.",
      label: 'Name of your first pet',
      name: 'mySpecialField',
      placeholder: 'Was it Fluffy?',
      value: 'Rex'
    };

    const { higTextArea, higContainer } = createHigTextArea(defaults);
    const container = document.createElement('div');
    const wrapper = mount(createOrionTextArea(defaults), {
      attachTo: container
    });
    setLabelForInputId(container);

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders a text field with udpated props', () => {
    const defaults = {
      name: 'mySpecialField'
    };

    const { higTextArea, higContainer } = createHigTextArea(defaults);
    const orionContainer = document.createElement('div');
    const wrapper = mount(createOrionTextArea(defaults), {
      attachTo: orionContainer
    });
    setLabelForInputId(orionContainer);

    const nextProps = {
      icon: 'assets',
      instructions: "Don't just do something, sit there.",
      label: 'Name of your first pet',
      name: 'updatedFieldName',
      placeholder: 'Was it Fluffy?',
      required: 'You really must fill this in.',
      value: 'Rex'
    };

    higTextArea.setInstructions(nextProps.instructions);
    higTextArea.setLabel(nextProps.label);
    higTextArea.setName(nextProps.name);
    higTextArea.setPlaceholder(nextProps.placeholder);
    higTextArea.required(nextProps.required);
    higTextArea.setValue(nextProps.value);

    wrapper.setProps(nextProps);

    expect(orionContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(orionContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  ['onBlur', 'onChange', 'onFocus', 'onInput'].forEach(eventName => {
    it(`sets event listeners for ${eventName} initially`, () => {
      const spy = jest.fn();
      const container = document.createElement('div');
      const wrapper = mount(createOrionTextArea({ [eventName]: spy }), {
        attachTo: container
      });
      setLabelForInputId(container);

      const instance = wrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(eventName);
      expect(disposeFunction).toBeDefined();
    });

    it(`sets event listeners for ${eventName} when updated`, () => {
      const spy = jest.fn();
      const container = document.createElement('div');
      const wrapper = mount(createOrionTextArea({}), {
        attachTo: container
      });
      setLabelForInputId(container);

      wrapper.setProps({ [eventName]: spy });

      const instance = wrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(eventName);
      expect(disposeFunction).toBeDefined();
    });
  });
});
