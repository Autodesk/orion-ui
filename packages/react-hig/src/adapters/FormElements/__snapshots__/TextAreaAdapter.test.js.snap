// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextAreaAdapter> renders a text field 1`] = `
"<div class=\\"hig__text-area\\">
    <textarea name=\\"mySpecialField\\" id=\\"1234\\" class=\\"hig__text-area__field hig__text-area__field--no-value\\" placeholder=\\"\\"></textarea>
    <!--LABEL-->
    <label for=\\"1234\\" class=\\"hig__text-area__label \\"></label>
    <!--INSTRUCTIONS-->
    
    <!--REQUIRED-NOTICE-->
</div>"
`;

exports[`<TextAreaAdapter> renders a text field with initial props 1`] = `
"<div class=\\"hig__text-area\\">
    <textarea name=\\"mySpecialField\\" id=\\"1234\\" class=\\"hig__text-area__field hig__text-area__field--has-value\\" placeholder=\\"Was it Fluffy?\\">Rex</textarea>
    <!--LABEL-->
    <label for=\\"1234\\" class=\\"hig__text-area__label hig__text-area__label--visible\\">Name of your first pet</label>
    <!--INSTRUCTIONS-->
    <p class=\\"hig__text-area__instructions\\">Don't just do something, sit there.</p>
    <!--REQUIRED-NOTICE-->
</div>"
`;

exports[`<TextAreaAdapter> renders a text field with udpated props 1`] = `
"<div class=\\"hig__text-area hig__text-area--required\\">
    <textarea name=\\"updatedFieldName\\" id=\\"1234\\" class=\\"hig__text-area__field hig__text-area__field--has-value\\" placeholder=\\"Was it Fluffy?\\">Rex</textarea>
    <!--LABEL-->
    <label for=\\"1234\\" class=\\"hig__text-area__label hig__text-area__label--visible\\">Name of your first pet</label>
    <p class=\\"hig__text-area__instructions\\">Don't just do something, sit there.</p><!--INSTRUCTIONS-->
    
    <p class=\\"hig__text-area__required-notice\\">You really must fill this in.</p><!--REQUIRED-NOTICE-->
</div>"
`;
